lang/LangPrimSource/PyrArrayPrimitives.cpp

```diff
@@ -2369,6 +2376,97 @@ int prArrayUnlace(struct VMGlobals *g, int numArgsPushed)
 	return errNone;
 }
 
+int prArrayPOpen(struct VMGlobals *g, int numArgsPushed)
+{
+	PyrObject *obj;
+	struct sc_process *process;
+	int err;
+
+	PyrSlot *a = g->sp - 1;
+	PyrSlot *b = g->sp;
+	
+	if (NotObj(a)) return errWrongType;
+
+	obj = slotRawObject(a);
+	if (!(slotRawInt(&obj->classptr->classFlags) & classHasIndexableInstances))
+		return errNotAnIndexableObject;
+		
+	if( obj->size < 1)
+		return errFailed;
+		
+	PyrSlot filenameSlot;
+		
+	getIndexedSlot(obj, &filenameSlot, 0);
+	if (!isKindOfSlot(&filenameSlot, class_string)) return errWrongType;
+	char *filename = (char*)malloc(slotRawObject(&filenameSlot)->size + 1);
+	err = slotStrVal(&filenameSlot, filename, slotRawObject(&filenameSlot)->size + 1);
+	if(err) {
+		free(filename);
+		return errFailed;
+	} 	
+	
+	char *argv[obj->size + 1];
+	
+	//execve expects first argument is command name
+	path p;
+	p /= filename;
+	const char *f = p.filename().string().c_str();
+	char *command = (char*)malloc(strlen(f));
+	strcpy(command, f );
+	argv[0] = command;
+	argv[obj->size] = NULL;
+		
+	if(obj->size > 1) {
+		for (int i=1; i<obj->size; ++i) {
+			PyrSlot argSlot;
+			getIndexedSlot(obj, &argSlot, i);
+			if (!isKindOfSlot(&argSlot, class_string)) return errWrongType;
+			char *arg = (char*)malloc(slotRawObject(&argSlot)->size + 1);
+			err = slotStrVal(&argSlot, arg, slotRawObject(&argSlot)->size + 1);
+			if(err) {
+				free(filename);
+				free(command);
+				free(arg);
+				if(i>1){
+					for (int j=1; j < (i-1); ++j) {
+						free(argv[j]);
+					}
+				}
+				return errFailed;
+			}
+			argv[i] = arg;
+		}
+	}
+
+#ifdef SC_IPHONE
+	SetInt(a, 0);
+	return errNone;
+#endif
+
+	process = (struct sc_process *)malloc(sizeof(struct sc_process));
+	process->stream = sc_popen_argv(filename, argv, &process->pid, "r");
+	setvbuf(process->stream, 0, _IONBF, 0);
+
+	process->postOutput = IsTrue(b);
+
+	free(filename);
+	for (int i=0; i<obj->size; ++i) {
+		free(argv[i]);
+	}
+
+	if(process->stream == NULL) {
+		free(process);
+		return errFailed;
+	}
+
+	thread thread(std::bind(string_popen_thread_func, process));
```

Comment:
it would be best to have a cleanup / deinit function (iirc i cleaned up the midi / hid primitives some time ago) ... but yes, a lot of this somehow dubious code is still in there ...

this all dates back to a time, when run loops were not yet invented :P